### PR TITLE
Type: Feat  - Add support for MariaDB connection string

### PR DESCRIPTION
**Pull request by tebrown**
### Description

Fixes: #490 

Add support for "mariadb" connection strings.  This is needed because sqlalchemy has a bit of special sauce for MariaDB vs MySQL, however the GIS interface between the two is identical.

Adding to `mariadb` to `geoalchemy2/admin/__init__.py` was necessary, but not sufficient.  I had to also add compiles for the mysql dialect (`geoalchemy2/admin/dialects/mysql.py`, `geoalchemy2/types/__init__.py`) that are for mariadb.  I thought about splitting them out, but because they are 100% identical, it didn't make sense. 

Lastly, I did add some tests. However, because MariadB and MySQL run on the same port, I didn't actually add any tests to it yet.  

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- [ ] Please include: `Fixes: #<issue number>` in the description if it solves an existing issue
	  (which must include a complete example of the issue).
	- [ ] Please include tests that fail with the `main` branch and pass with the provided fix.
- [X] A new feature implementation or update an existing feature
	- [X] Please include: `Fixes: #<issue number>` in the description if it solves an existing issue
	  (which must include a complete example of the feature).
	- [N/A] Please include tests that cover every lines of the new/updated feature.
	- [N/A] Please update the documentation to describe the new/updated feature.

### CODE REVIEW

File: geoalchemy2/types/__init__.py [hunk 1]
```diff
@@ -196,6 +197,20 @@ def check_ctor_args(geometry_type, srid, dimension, use_typmod, nullable):
...
+
+    if not self.nullable or self.spatial_index:
+        spec += " NOT NULL"
+    if self.srid > 0:
+        spec += " SRID %d" % self.srid
+    return spec
+
+
 @compiles(_GISType, "mysql")
 def get_col_spec(self, *args, **kwargs):
```
Comment (adrien-berchet) [inline geoalchemy2/types/__init__.py:215]:
```suggestion
@compiles(_GISType, "mysql")
@compiles(_GISType, "mariadb")
def get_col_spec(self, *args, **kwargs):
```
I think these decorators can also be combined, can't they?

File: geoalchemy2/admin/dialects/mysql.py [hunk 2]
```diff
@@ -192,3 +192,22 @@ def _MySQL_ST_GeomFromWKB(element, compiler, **kw):
 @compiles(functions.ST_GeomFromEWKB, "mysql")  # type: ignore
 def _MySQL_ST_GeomFromEWKB(element, compiler, **kw):
     return _compile_GeomFromWKB_MySql(element, compiler, **kw)
+
+@compiles(functions.ST_GeomFromText, "mariadb")  # type: ignore
```
Comment (adrien-berchet) [inline geoalchemy2/admin/dialects/mysql.py:196]: I think that it's possible to combine these decorators like:
```python
@compiles(functions.ST_GeomFromText, "mysql")  # type: ignore
@compiles(functions.ST_GeomFromText, "mariadb")  # type: ignore
def _MySQL_ST_GeomFromText(element, compiler, **kw):
    return _compile_GeomFromText_MySql(element, compiler, **kw)
```
Can you test it please?

**Review by adrien-berchet — COMMENTED** [review]
Thank you very much for this PR @tebrown !
I'm ok with the principle of this PR but I think it can be simpler, see the 2 comments.

**Comment by adrien-berchet** [comment]
Ah and also, could you write something in the docs about the compatibility with MariaDB please? (for know it should be clearly mentioned that it's experimental)

File: geoalchemy2/types/__init__.py (hunk 1 above)
Comment (tebrown) [reply geoalchemy2/types/__init__.py:215]: Of course!  I should have thought of that. Let me test it and see if it works!

File: geoalchemy2/admin/dialects/mysql.py (hunk 2 above)
Comment (tebrown) [reply geoalchemy2/admin/dialects/mysql.py:196]: Yup

File: geoalchemy2/types/__init__.py (hunk 1 above)
Comment (adrien-berchet) [reply geoalchemy2/types/__init__.py]: Thanks! That's very nice like this :)

**Comment by adrien-berchet** [comment]
Looks good!
Let's see what the CI thinks about it :)
I guess it will complain if we don't set up a MariaDB instance in CI?

**Comment by tebrown** [comment]
Tested and all these changes work. Thank you for the feedback!

**Comment by tebrown** [comment]
> Looks good! Let's see what the CI thinks about it :) I guess it will complain if we don't set up a MariaDB instance in CI?

TBD!  If it does, I can see about updating that.  MariaDB is free, but I am not 100% sure how to handle the port conflicts.  I will noodle on that.

**Comment by tebrown** [comment]
It looks like it failed on codespell.  I'll see if I can fix that.

**Comment by tebrown** [comment]
It actually caught a bug!  Neat!

**Comment by tebrown** [comment]
Looks like I'll be writing some tests!

**Comment by tebrown** [comment]
@adrien-berchet I think I might be stuck.  I see the mapping for the `ST_AsWEKB' to 'ST_AsBinary' in `admin/dialects/mysql.py`, and I believe that this should be responsible for making the query SQLAlchemy/GeoAlchemy generates correct.  However, this code:

```python
async def test():
    async with async_session() as session:
        await session.execute(select(LocationAlert).where(LocationAlert.psap == 'psap'))
```

Generates the following SQL:
```log
sqlalchemy.exc.OperationalError: (asyncmy.errors.OperationalError) (1305, 'FUNCTION orthoplex.ST_AsEWKB does not exist')
[SQL: SELECT location_alerts.psap, location_alerts.call_type, location_alerts.dst_agency, location_alerts.units, ST_AsEWKB(location_alerts.mapped_geom) AS mapped_geom, location_alerts.id 
FROM location_alerts 
WHERE location_alerts.psap = %s]
[parameters: ('psap',)]
(Background on this error at: https://sqlalche.me/e/20/e3q8)
```

If I manually take that SQL and change it to `ST_AsBinary`, it works great.  But for some reason, this is generating the wrong function name.  I went through every place that I could find that relates to MySQL, and it all looks kosher.  What is even worse, is that it is properly generating the tables, so it MUST be partially working.

**Comment by tebrown** [comment]
Well, I spoke too soon. I missed a spot.  I am the worst!

**Comment by adrien-berchet** [comment]
Ahah nice catch :⁠-⁠)

**Comment by tebrown** [comment]
Ok, I ran the tests locally this time, and I see errors for sqlalatest adn sqla14...however I didn't change those, and it is probably unlikely I impacted them.

```
ERROR:   py37-sqla14: commands failed
ERROR:   py37-sqlalatest: commands failed
ERROR:   py38-sqla14: commands failed
ERROR:   py38-sqlalatest: commands failed
ERROR:   py39-sqla14: commands failed
ERROR:   py39-sqlalatest: commands failed
ERROR:   py310-sqla14: commands failed
ERROR:   py310-sqlalatest: commands failed
ERROR:   py311-sqla14: commands failed
ERROR:   py311-sqlalatest: commands failed
ERROR:   pypy3-sqla14: commands failed
ERROR:   pypy3-sqlalatest: commands failed
```

File: tests/test_functional.py [hunk 3]
```diff
@@ -431,9 +431,9 @@ def test_WKBElement(self, session, Lake, setup_tables, dialect_name):
         srid = session.execute(lake.geom.ST_SRID()).scalar()
         assert srid == 4326
 
-    @test_only_with_dialects("postgresql", "mysql", "sqlite-spatialite3", "sqlite-spatialite4")
+    @test_only_with_dialects("postgresql", "mysql", "sqlite-spatialite3", "sqlite-spatialite4"]:
```
Comment (adrien-berchet) [inline tests/test_functional.py:434]:
```suggestion
    @test_only_with_dialects("postgresql", "mysql", "sqlite-spatialite3", "sqlite-spatialite4")
```

File: tests/test_functional.py [hunk 4]
```diff
@@ -693,7 +693,7 @@ def test_ST_AsGeoJson(self, session, Lake, setup_one_lake):
         assert loads(r1_func) == {"type": "LineString", "coordinates": [[0, 0], [1, 1]]}
 
     @skip_case_insensitivity()
-    @test_only_with_dialects("postgresql", "mysql", "sqlite-spatialite3", "sqlite-spatialite4")
+    @test_only_with_dialects("postgresql", "mysql", "sqlite-spatialite3", "sqlite-spatialite4"):
```
Comment (adrien-berchet) [inline tests/test_functional.py:696]:
```suggestion
    @test_only_with_dialects("postgresql", "mysql", "sqlite-spatialite3", "sqlite-spatialite4")
```

**Review by adrien-berchet — COMMENTED** [review]
If we don't add tests for MariaDB in this PR (we can suppose it just works as MySQL) then you should revert all your changes in the tests.

File: tests/test_functional.py (hunk 3 above)
Comment (tebrown) [reply tests/test_functional.py:434]: Should be all set!

**Comment by adrien-berchet** [comment]
I reverted the tests so for me it's ok to merge if the CI passes.

**Comment by adrien-berchet** [comment]
Thanks for this work @tebrown !
I'm going to push a new release so you can use it in your workflow.

**Pull request merged**